### ISSUE TITLE
[$1000] IOU - Request money flow breaks on second time after the 1st one paid

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16105]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Login to an account A
2. Complete a request money with Account B
3. Pay via Expensify as Account B
4. Request money from Account B logged in as Account A

## Expected Result:
User is able to request money 2nd time

## Actual Result:
Error message occurs as "Unexpected error requesting money. Please try again later". 

## Workaround:
User has to log out and login back to request again

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.87-0
**Reproducible in staging?:** Needs reproduction
**Reproducible in production?:** Needs reproduction

**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
<img width="891" alt="image (1)" src="https://user-images.githubusercontent.com/43996225/226212178-55f210e5-2fe1-4b0e-8f2b-931622d242ea.png">

<img src="https://user-images.githubusercontent.com/43996225/226212179-5e001d9d-7d17-46e2-919f-aa804c3866ea.png" width="40%" height="40%">

**Expensify/Expensify Issue URL:**
**Issue reported by:** @quinthar 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1674068986019339

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01a79e6662f659095e</li>
        <li>Upwork Job ID: 1677085388442824704</li>
        <li>Last Price Increase: 2023-07-06</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @sophiepintoraetz (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by sophiepintoraetz** [comment]
Tomorrow.

**Comment by sophiepintoraetz** [comment]
Hopefully have some time today - now that I'm recovering!

**Comment by quinthar** [comment]
This is 100% reproducible.  I just got it again today.

<img width="467" alt="image" src="https://user-images.githubusercontent.com/575186/227623141-c218991d-3078-424b-9ec8-c06eebeaa7c7.png">

**Comment by sophiepintoraetz** [comment]
I also don't have the ability to reproduce this given I don't have the ability to set up a wallet (being outside the US) - but I see that Alberto [has been able to reproduce this](https://expensify.slack.com/archives/C049HHMV9SM/p1674139515772709?thread_ts=1674068986.019339&cid=C049HHMV9SM), so swinging it over to Engineering.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @MariaHCD (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MariaHCD** [comment]
Able to reproduce the bug on dev, looks like a unique constraints violation:

<img width="1789" alt="Screenshot 2023-03-28 at 12 56 17 PM" src="https://user-images.githubusercontent.com/12268372/228243205-d4ad1a86-24f1-4c31-a4fe-9651e7809ed5.png">

```
2023-03-28T12:58:33.881339+00:00 expensidev2004 php-fpm: 6UxSxh /api.php test@pendingwallet.com !ecash1.2.90-6! ?api? [info] Bedrock\Client - Request finished ~~ host: 'auth' command: 'CreateIOUTransaction' jsonCode: '400 Unique Constraints Violation' duration: '6' net: '2.777' wait: '1.547' proc: '1.676' commitCount: '26485'

2023-03-28T12:58:33.883389+00:00 expensidev2004 bedrock: 6UxSxh test@pendingwallet.com (libstuff.cpp:2656) SQuery [socket112] [warn] 'read/write transaction', query failed with error #19 (UNIQUE constraint failed: reports.reportID): INSERT INTO reports ( reportID, created, state, accountID, reportName, currency ) VALUES ( 4579177482072177, '2023-03-28 12:58:33', 0, 137, 'IOU', 'USD' );
```

Seems like the iouReportID is the same in both requests:

<img width="323" alt="Screenshot 2023-03-28 at 12 57 16 PM" src="https://user-images.githubusercontent.com/12268372/228243461-a58d71f7-b6a1-444a-8f38-c32951bcd04f.png">


<img width="358" alt="Screenshot 2023-03-28 at 12 57 21 PM" src="https://user-images.githubusercontent.com/12268372/228243458-ae44ba4b-ee69-4417-9f4a-e843d92de7db.png">

**Comment by MariaHCD** [comment]
While the fix works for web and desktop, iOS still seems to fail with the same error. Looking into this.

**Comment by MelvinBot** [comment]
@MariaHCD, @sophiepintoraetz Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MariaHCD** [comment]
I'm adding tests to the PR but ran into issues with the iOS build so I'm downgrading XCode to 14.2 to fix that. https://expensify.slack.com/archives/C01GTK53T8Q/p1680773983844719

**Comment by MariaHCD** [comment]
https://github.com/Expensify/App/pull/16624 is out for review and is just pending on tests from the PR reviewers on native Android and iOS.

**Comment by sophiepintoraetz** [comment]
Hey @MariaHCD - what do we need to do to get this one closed out?

**Comment by MariaHCD** [comment]
The [PR](https://github.com/Expensify/App/pull/16624) was just merged 🙌🏼

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.0-2 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/16624


If no regressions arise, payment will be issued on 2023-04-24. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@MariaHCD] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@MariaHCD] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@MariaHCD] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@sophiepintoraetz] Determine if we should create a regression test for this bug.
- [ ] [@MariaHCD] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@sophiepintoraetz] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by sophiepintoraetz** [comment]
PR merged and internal so no payments due here - closing!

**Comment by quinthar** [comment]
This is not fixed.

**Comment by sophiepintoraetz** [comment]
Discussing reproducing this bug in NewDot - will try to loop back and narrow down where this error is occurring (all accounts or specific ones)

**Comment by sophiepintoraetz** [comment]
@MariaHCD - what are the next steps for us here? I don't have access to reproduce this bug but did put a[ call out](https://expensify.slack.com/archives/C01SKUP7QR0/p1683506857210309) to see if anyone else uses the gold wallet feature?

**Comment by MariaHCD** [comment]
Thanks, @sophiepintoraetz. I'm going to try to reproduce this locally and see if I can figure out what's going wrong.

**Comment by sophiepintoraetz** [comment]
Puneet mentioned that he has a gold wallet!

**Comment by MariaHCD** [comment]
I've been prioritizing some dailies but will get back to this next week!

Also, I don't think anyone external needs to get paid here so I updated the issue title so it isn't greyed out on my K2 dashboard

**Comment by MariaHCD** [comment]
There have been quite a few improvements/changes to the IOU flow. @quinthar Are you still facing this issue?

**Comment by quinthar** [comment]
Just just reproduced it today.

**Comment by MariaHCD** [comment]
As a next step, I’m going to look into this locally to see if I can reproduce the issue.

**Comment by MariaHCD** [comment]
Still able to reproduce the bug locally:

<img width="1787" alt="Screenshot 2023-06-08 at 2 32 35 PM" src="https://github.com/Expensify/App/assets/12268372/3837d70d-96cb-410c-8b7d-c99002a3cd65">

Looks like the error thrown by `PayWithWallet` was `406 Chat report with these participants already exists with different ID`:

```
2023-06-08T10:29:45.080887+00:00 expensidev2004 bedrock: VDAW0C test@activatedwallet.com (libstuff.cpp:151) SException [socket87] [info] Throwing exception with message: '406 Chat report with these participants already exists with different ID' from auth/command/PayWithWallet.cpp:170
2023-06-08T10:29:45.091854+00:00 expensidev2004 bedrock: VDAW0C test@activatedwallet.com (BedrockCore.cpp:285) _handleCommandException [socket87] [info] Error processing command 'PayWithWallet' (406 Chat report with these participants already exists with different ID), ignoring.
```

Logs from @quinthar's last request [here](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%227d33caec6f8feb43-SEA%22)+AND+timestamp:[2023-06-06T20:36:45.616Z+TO+2023-06-06T22:36:45.616Z]&index=logs_expensify-022688).

This error is being thrown in auth here: https://github.com/Expensify/Auth/blob/107c4c9951faa5f38369eef48b7a22e86d5c4410/auth/command/PayWithWallet.cpp#L166-L171

**Comment by maddylewis** [comment]
updating to daily based on [this SO](https://stackoverflowteams.com/c/expensify/questions/8208#:~:text=Make%20sure%20the%20GH%20is%20labelled%20with%20%60daily%60%20(unless%20it%20has%20%22Hold%22%20in%20the%20title)).

**Comment by MariaHCD** [comment]
Looking at this again today and I'm noticing it isn't consistently reproducible when I'm testing locally. 

@mountiny @luacmartins any idea why the app would sometimes send the wrong `chatReportID` in PayMoneyRequestWithWallet? (Full logs [here](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%227d33caec6f8feb43-SEA%22)+AND+timestamp:%5B2023-06-06T20:36:45.616Z+TO+2023-06-06T22:36:45.616Z%5D&index=logs_expensify-022688)) It seems to sometimes send the iouReportID as the chatReportID.

![Screenshot 2023-06-13 at 3 40 59 PM](https://github.com/Expensify/App/assets/12268372/7205a081-efa6-4967-8003-d63cff2f8712)

Looked at the usages of `IOU.payMoneyRequest` and I don't see how the wrong reportID could be passed.

**Comment by mountiny** [comment]
@MariaHCD I had the same investigation results on another issue https://github.com/Expensify/App/issues/19960

I am not sure though same as you, gotta look more into this

**Comment by luacmartins** [comment]
I see what's happening here. Notice that the payer has the same `chatReportID` and `reportID`in the iouReport Onyx data:

![Screenshot 2023-06-13 at 12 04 46 PM](https://github.com/Expensify/App/assets/22219519/acc8088a-f649-47c4-83d9-2ebf7cab3c8e)

This happens because:
1. The IOU report is opened when we tap the `reportPreview`
2. The `iouReportID` is passed as the `reportID` param in [ReadOldestAction](https://github.com/Expensify/Web-Expensify/blob/4c189319a34ccf6f24e22fb42330a4c1807690bf/lib/ReportAPI.php#L4580) 
3. We pass that `reportID` as `chatReportID` param in [ReportAPI::getSettledIOUReports](https://github.com/Expensify/Web-Expensify/blob/4c189319a34ccf6f24e22fb42330a4c1807690bf/lib/ReportAPI.php#L4597) 
4. An Onyx update with the incorrect `chatReportID` is sent to the payer

I think we can get rid of that call to `ReportAPI::getSettledIOUReports` in `readOldestAction` entirely, since we're now fetching those in [OpenReport](https://github.com/Expensify/Web-Expensify/blob/4c189319a34ccf6f24e22fb42330a4c1807690bf/lib/ReportAPI.php#L4753) (passing the correct chatReportID)

I put up a PR to fix it [here](https://github.com/Expensify/Web-Expensify/pull/37809)

**Comment by MariaHCD** [comment]
Thanks @luacmartins! That fix makes sense to me.

**Comment by sophiepintoraetz** [comment]
Waiting on @quinthar to [confirm](https://expensify.slack.com/archives/C01GTK53T8Q/p1687821393880859?thread_ts=1687191859.538319&cid=C01GTK53T8Q) whether they are still experiencing the problem.

**Comment by luacmartins** [comment]
This is still an issue, but I think the underlying root cause has changed since we deployed our last fix. From the [logs](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%227dd63b87bfddacd7-ATL%22)+AND+timestamp:%5B2023-06-26T13:45:11.162Z+TO+2023-06-26T15:45:11.162Z%5D&index=logs_expensify-022988), we're hitting [this line](https://github.com/Expensify/Auth/blob/50520ee1d81cf9261349de5e7224765ce4aac19e/auth/lib/Report.cpp#L3903) which means we failed to retrieve an open IOU report for those participants. That makes sense since the reportID 3268887268846589 was originally created on Jun 22nd and paid on Jun 23rd. However, App still passed it as the active IOU report for the new request on Jun 26.

I'm still not sure why though. The only way I found to reproduce this issue was:

1. userA requests money from userB
2. userA goes offline
3. userB pays the request
4. userA requests money
5. userA goes online

In this case, userA would send the already paid reportID as param and fail since that report had already been paid and was already taken in the DB. So it's possible that this was a product of a missed pusher update/flaky network.

**Comment by sophiepintoraetz** [comment]
@luacmartins - do you have an update on this?

**Comment by luacmartins** [comment]
I DMed David's assistant and we're in the process of testing this further. 

1. [Logs](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%227df853086e802819-SEA%22)+AND+timestamp:[2023-06-30T17:03:09.537Z+TO+2023-06-30T19:03:09.537Z]&index=logs_expensify-023064) for her request
2. [Logs](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%227e1ab5a13a97bef4-SJC%22)+AND+timestamp:[2023-07-04T21:12:24.118Z+TO+2023-07-04T23:12:24.118Z]&index=logs_expensify-023135) for me paying the request

Currently waiting for her to make another request

**Comment by luacmartins** [comment]
Ok, I just got the 2nd request and it failed. Here are the [logs](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%227e22e606be253096-SEA%22)+AND+timestamp:[2023-07-05T21:03:32.320Z+TO+2023-07-05T23:03:32.320Z]&index=logs_expensify-023154).

![Screenshot 2023-07-05 at 4 10 07 PM](https://github.com/Expensify/App/assets/22219519/cbf1fe4a-d0e3-4130-b4a9-326a9bec7d83)

So it seems like the problem is in the client since we're sending the same `iouReportID` and `reportPreviewActionID` even after paying the IOU.


**1st request**
```
iouReportID: 8247186926908861
chatReportID: 4776780779232861
transactionID: 8493050559591231243
createdChatReportActionID: 0
createdIOUReportActionID: 3707095983714300591
reportPreviewActionID: 2957006752648145465
```

**2nd request**
```
iouReportID: 8247186926908861
chatReportID: 4776780779232861
transactionID: 4187278194600692682
createdChatReportActionID: 0
createdIOUReportActionID: 0
reportPreviewActionID: 2957006752648145465
```

I'll investigate this further tomorrow.

**Comment by mountiny** [comment]
Should we maybe add more checks to [this line](https://github.com/Expensify/App/blob/7e908b17565e7bf659309f4b4d95141e70094ab9/src/libs/actions/IOU.js#L339) as there is a chance the chatReport data was not updated (pusher missed) and so the iouReportID is the one already paid. We can then also check is the iouReport of that ID is settled already and if yes, then also consider this as new IOU report flow.

**Comment by luacmartins** [comment]
Well, we ruled out missing the pusher update by making sure we're online when sending the requests. I think adding more checks in [this line](https://github.com/Expensify/App/blob/7e908b17565e7bf659309f4b4d95141e70094ab9/src/libs/actions/IOU.js#L339) sounds good, but I'd like to reproduce this first to make sure that fixes the problem. I'm still having issues to reproduce this locally though.

**Comment by luacmartins** [comment]
The error above happens because we're trying to get the `reportID` from the route [here](https://github.com/Expensify/App/blob/34b792a07d76fb5064ed7be78b907669dffc1dd6/src/pages/iou/steps/MoneyRequestConfirmPage.js#L235), but that's not available (see image below) so we default to `''` and end up returning all reports stored in Onyx!

![Screenshot 2023-07-06 at 11 49 35 AM](https://github.com/Expensify/App/assets/22219519/d75c74bb-d45c-4843-a861-de9d67413dc8)

**Comment by luacmartins** [comment]
The only way I can reproduce this issue locally is by following [these steps](https://github.com/Expensify/App/issues/16105#issuecomment-1610102668). However, we ruled those out in our latest round of testing. I'm gonna make this external to get more eyes on it and maybe get someone to reproduce the issue.

**Comment by luacmartins** [comment]
>It is indeed because we don't receive the Pusher update to clear the iouReportID while offline or while the app is not open.

I think this can only be the case if we are **online** but miss the pusher update for some reason. Otherwise, `OpenApp` or `ReconnectApp` would fetch the updated chat report when the user comes back online and this wouldn't be an issue.

**Comment by luacmartins** [comment]
PR [here](https://github.com/Expensify/Web-Expensify/pull/38128)

**Comment by luacmartins** [comment]
@bernhardoj gonna assign you to the issue for payment since you helped triage it

**Comment by mountiny** [comment]
Yep I think eventually we would like to do 2. Should we do it now?

**Comment by luacmartins** [comment]
Yea, I think we should.

**Comment by luacmartins** [comment]
PR will be deployed to prod today? so we can retest this then.

**Comment by luacmartins** [comment]
I tested this in prod today and it seems to be fixed! @quinthar let me know if you're still having issues with failed money requests.

**Comment by sophiepintoraetz** [comment]
@luacmartins - in terms of a PR review - is this the standard amount of $1000 for both @bernhardoj and @sobitneupane?

**Comment by luacmartins** [comment]
@sophiepintoraetz we'd pay $1000 to @bernhardoj for the proposal and $1000 to @sobitneupane for the PR review.

**Comment by sophiepintoraetz** [comment]
All right - offers sent! 
@luacmartins - we could also check in with Emma to see? Otherwise, @quinthar - if you're still encountering an issue with this, reopen the GH! Thanks Carlos for pushing this across the finish line 🤍

**Comment by sobitneupane** [comment]
@sophiepintoraetz I didn't review the PR. So, no payment for me.

**Comment by sophiepintoraetz** [comment]
All set here.
